src/indexable.cr

```diff
@@ -479,7 +479,7 @@ module Indexable(T)
   # ```
   # [1, 2, 3, 1, 2, 3].index(2, offset: 2) # => 4
   # ```
-  def index(object, offset : Int = 0)
+  def index(object, offset : Int = 0) : Int32 | UInt32 | UInt64 | Nil
```

Comment:
no `Int64`?

Reply:
This should be `Int32 | Nil` or `Int32?`

Reply:
does crystal upcast implicitly? I think it's used at least once in a context expecting each of these types. Will check again.

Reply:
The return type is the same as `offset` (or `Nil`), because it's derived from that value. So it should probably just be specified as `Int?`.

Reply:
The return type is just checked but not used. Well, it's used in one specific case where a type can't be determined because of recursion. So it's better to put the more strict type, which is `Int32?`.

Reply:
@straight-shoota that was my first attempt, but the compiler rejects `Int?` (because it's unsized I imagine)

Reply:
I'd take a step back and first ask: Should it be mandatory for `Indexable(T)` to have a Int32 size? Or should it be up to the implementing T to decide whatever type its size has? If the latter, I'd vote for removing the return type signature here altogether.

Reply:
I'm surprised there isn't a `Size` integer type, which Indexable could use. Or the index type could be a parameter of `Indexable` itself.

Reply:
There is a discussion about it: https://github.com/crystal-lang/crystal/issues/8111

But for now Int32 is the default integer type that should be used everywhere

Reply:
I am not sure this is resolved I think this should be `Int32?`. I would not care as much if this was not locking down public APIs and could cause breaking changes in the future

Reply:
putting `Int32?` makes some specs fail because they use `index` and expect a `UInt64?` or `UInt32?` (which are otherwise properly inferred).

As it seems there's no clean way to say something like `def index(…) : U? forall U < Int`, I removed the annotation.